sklearn/datasets/openml.py

```diff
@@ -0,0 +1,516 @@
+import gzip
+import json
+import os
+import shutil
+from os.path import join, exists
+from warnings import warn
+
+try:
+    # Python 3+
+    from urllib.request import urlopen
+except ImportError:
+    # Python 2
+    from urllib2 import urlopen
+
+
+import numpy as np
+import scipy.sparse
+
+from sklearn.externals import arff
+from .base import get_data_home
+from ..externals.six import string_types, PY2
+from ..externals.six.moves.urllib.error import HTTPError
+from ..utils import Bunch
+
+__all__ = ['fetch_openml']
+
+_OPENML_PREFIX = "https://openml.org/"
+_SEARCH_NAME = "api/v1/json/data/list/data_name/{}/limit/1"
+_DATA_INFO = "api/v1/json/data/{}"
+_DATA_FEATURES = "api/v1/json/data/features/{}"
+_DATA_FILE = "data/v1/download/{}"
+
+
+def _open_openml_url(openml_path, data_home):
+    """
+    Returns a resource from OpenML.org. Caches it to data_home if required.
+
+    Parameters
+    ----------
+    openml_path : str
+        OpenML URL that will be accessed. This will be prefixes with
+        _OPENML_PREFIX
+
+    data_home : str
+        Directory to which the files will be cached. If None, no caching will
+        be applied.
+
+    Returns
+    -------
+    result : stream
+        A stream to the OpenML resource
+    """
+    if data_home is None:
+        return urlopen(_OPENML_PREFIX + openml_path)
+    local_path = os.path.join(data_home, 'openml.org', openml_path + ".gz")
+    if not os.path.exists(local_path):
+        try:
+            os.makedirs(os.path.dirname(local_path))
```

Comment:
use `if not os.path.exists(os.path.dirname(local_path))` instead of try-except. Otherwise you'll catch other OSErrors.

Reply:
Not entirely sure. When running this code on a cluster, it can happen that several processes arrive at the same time at this code-line. When they all try to make the directory, only one will succeed and the rest will fail, resulting in a crash. Python 3.X has the argument 'exists_ok' to circumvent this problem, but in Python 2.7 it is common practice to use the try except struct. What do you think?

Reply:
Okay. You're right that race conditions are an issue. At least you'll get an error below if it continues to not exist.